2020-01-Winter/5_graphs/TelephoneStrangers/TelephoneStrangers.cpp

```diff
@@ -0,0 +1,81 @@
+#include <iostream>
+#include <map>
+#include <vector>
+#include <utility>
+#include <set>
+#include "Graph.h"
+#include "Vertex.h"
+#include "Edge.h"
+
+std::pair<std::string, std::string> Helper(std::set<Vertex *> InCallHistory, Vertex *Curr)
+{
+    for (auto I : Curr->returnCallsMade())
+    {
+        auto Temp = InCallHistory.find(I->returnCalled()); 
+        if(Temp == InCallHistory.end())
+        {
+            continue; // update this
+        }
+
+        InCallHistory.erase(I->returnCalled());
+    }
+
+    std::vector<Vertex *> Temp;
+    for (auto I : InCallHistory)
+    {
+        Temp.push_back(I);
+    }
+    auto Foo = std::make_pair(Temp[0]->returnNumber(), Temp[1]->returnNumber());
+    return Foo;
+}
+
+std::vector<std::pair<std::string, std::string>>
+FindStrangers(Graph *PhoneHistory)
+{
+    std::vector<std::pair<std::string, std::string>> StrangerPairs;
+    std::set<Vertex *> PrevCallStack;
+    auto Numbs = PhoneHistory->returnPhoneNumbers();
+
+    for (auto I : Numbs)
+    {
+        PrevCallStack.insert(I.second);
+    }
+
+    for (auto I : Numbs)
+    {
+        if(I.second->returnCallsMade().empty())
+        {
+            continue;
+        }
+        StrangerPairs.push_back(Helper(PrevCallStack, I.second));
+    }
+
+    return StrangerPairs;
+}
+
+
+void test1()
+{
+    auto G1 = new Graph();
+    std::map<std::string, std::vector<std::string>> CallHistory;
+    std::vector<std::string> VectList = {};
+    CallHistory.insert({"206", VectList});
+    VectList = {"425"};
+    CallHistory.insert({"940", VectList});
+    VectList = {"940"};
+    CallHistory.insert({"425", VectList});
+    G1->addPhone(CallHistory);
+    FindStrangers(G1);
+}
+
+void test2()
+{
+    // ? Test the second case in problem
+}
```

Comment:
Adding second test case introduces a segfault and I haven't figured out why just yet. Will look at it more later tonight & tomorrow.

Also, bless your heart for writing this in C++. _Python might actually change your life, pro tip_

Reply:
A python solution: 

```python
def find_strangers(call_records):
    strangers = []
    for num in call_records:
        for other_num in call_records:
            # edge case since we're iterating through keys twice
            if num == other_num: continue
            if other_num not in call_records[num]:
                # no edge exists between the vertices
                pair = tuple(sorted([num, other_num]))
                if pair not in strangers: strangers.append(pair)
    return strangers

def tests():
    t1 = { '206': [], '940': ['425'], '425': ['940'] }
    print('Test case 1:', t1)
    r1 = find_strangers(t1)
    print('Test 1 result:', r1)
    assert(len(r1) == 2)
    assert(('206', '425') in r1)
    assert(('206', '940') in r1)
    t2 = { '1': ['2', '3', '4'], '2': ['1', '4'], '3': ['1'], '4': ['1', '2'] }
    print('Test case 2:', t2)
    r2 = find_strangers(t2)
    print('Test 2 result:', r2)
    assert(len(r2) == 2)
    assert(('2', '3') in r2)
    assert(('3', '4') in r2)

if __name__=="__main__":
    tests()
```

That's all to it. Python is my favorite language for reasons like this. C++ is powerful in its granularity, but Python is always a pleasure to write.

Reply:
> Adding second test case introduces a segfault and I haven't figured out why just yet. Will look at it more later tonight & tomorrow.

@Erica-Fer recommend looking at c01df446b4c02c5d8ae59a602824698b0f984fb4; the 2nd example caused indexing into a vector that was empty and threw a segfault for an invalid memory read